src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,249 @@
+
+          vehicle.addFuel(providedEnergy)
+          val currentSession = ChargingSession(providedEnergy, chargingDuration)
+          val newTotalSession = totalChargingSession.combine(currentSession)
+
+          endRefuelSessionTriggerMaybe(
+            vehicle,
+            tick,
+            maxChargeDuration,
+            currentSession,
+            newTotalSession
+          ).map { endRefuelSession =>
+              vehiclesToCharge.remove(vehicleId)
+              ScheduleTrigger(endRefuelSession, agent)
+            }
+            .orElse {
+              vehiclesToCharge.update(vehicleId, ChargingVehicle(vehicle, agent, newTotalSession, currentSession))
+              None
+            }
+
+        case (id, energy) if energy <= 0 =>
+          log.warning(
+            "Vehicle {}  (primaryFuelLevel = {}) requires energy {} - which is less or equals zero",
+            vehicles(id),
+            vehicles(id).primaryFuelLevelInJoules,
+            energy
+          )
+          None
+      }.toVector
+
+      sender ! CompletionNotice(
+        triggerId,
+        if (tick < endOfSimulationTime)
+          scheduleTriggers :+ ScheduleTrigger(PlanningTimeOutTrigger(nextTick), self)
+        else {
+          // if we still have a BEV/PHEV that is connected to a charging point,
+          // we assume that they will charge until the end of the simulation and throwing events accordingly
+          val completeTriggers = scheduleTriggers ++ vehiclesToCharge.map {
+            case (_, cv) =>
+              ScheduleTrigger(
+                EndRefuelSessionTrigger(
+                  tick,
+                  cv.vehicle.getChargerConnectedTick(),
+                  cv.totalChargingSession.energy,
+                  cv.vehicle
+                ),
+                cv.agent
+              )
+          }
+          vehiclesToCharge.clear()
+          completeTriggers
+        }
+      )
+  }
+
+  private def replanHorizon(tick: Int): (Int, Map[Id[BeamVehicle], Double]) = {
+    val requiredPower = sitePowerManager.getPowerOverPlanningHorizon(vehicles)
+
+    val (bounds, nextTick) = if (gridConnectionEnabled) {
+      powerController.publishPowerOverPlanningHorizon(requiredPower, tick)
+      powerController.obtainPowerPhysicalBounds(tick)
+    } else {
+      powerController.defaultPowerPhysicalBounds(tick)
+    }
+    val requiredEnergyPerVehicle = sitePowerManager.replanHorizonAndGetChargingPlanPerVehicle(bounds, vehicles)
+
+    if (requiredEnergyPerVehicle.nonEmpty)
+      log.info("Required energy per vehicle: {}", requiredEnergyPerVehicle.mkString(","))
+
+    (nextTick, requiredEnergyPerVehicle)
+  }
+
+  private def endRefuelSessionTriggerMaybe(
+    vehicle: BeamVehicle,
+    tick: Int,
+    maxChargeDuration: Long,
+    currentSession: ChargingSession,
+    totalSession: ChargingSession
+  ): Option[EndRefuelSessionTrigger] = {
+    if (vehicle.primaryFuelLevelInJoules >= vehicle.beamVehicleType.primaryFuelCapacityInJoule) {
```

Comment:
I would assume if it is more than the capacity then this is a bug and should result in an error and reset back to capacity?

Reply:
Good point!